diff --git a/lib/commands/set/version.js b/lib/commands/set/version.js
index 72f9dce6e7a2835da6f1bf42c9b7649a7e58a1d5..f4e5fc22664c8e0276984583ade04c1f38b6ad67 100644
--- a/lib/commands/set/version.js
+++ b/lib/commands/set/version.js
@@ -83,7 +83,7 @@ class SetVersionCommand extends cli_1.BaseCommand {
     }
 }
 SetVersionCommand.paths = [
-    [`set`, `version`],
+    [`set`, `version`, `original`],
 ];
 SetVersionCommand.usage = clipanion_1.Command.Usage({
     description: `lock the Yarn version used by the project`,
