diff --git a/lib/resolveConfig.js b/lib/resolveConfig.js
index cb4ab0bde520aff5a75f1c6ab7438ba15188915f..5ce7c70f42e9f06712b47e03b958ec303081038b 100644
--- a/lib/resolveConfig.js
+++ b/lib/resolveConfig.js
@@ -4,10 +4,9 @@ import { createRequire } from 'node:module'
  * require() does not exist for ESM, so we must create it to use require.resolve().
  * @see https://nodejs.org/api/module.html#modulecreaterequirefilename
  */
-const require = createRequire(import.meta.url)
-
 export function resolveConfig(configPath) {
   try {
+    const require = createRequire(import.meta.url)
     return require.resolve(configPath)
   } catch {
     return configPath
